artichoke-backend/src/extn/core/string/trampoline.rs

```diff
@@ -586,10 +586,24 @@ pub fn byteindex(
     offset: Option<Value>,
 ) -> Result<Value, Error> {
     #[cfg(feature = "core-regexp")]
-    if let Ok(_pattern) = unsafe { Regexp::unbox_from_value(&mut substring, interp) } {
-        return Err(NotImplementedError::from("String#byteindex with Regexp pattern").into());
-    }
     let s = unsafe { super::String::unbox_from_value(&mut value, interp)? };
+    if let Ok(pattern) = unsafe { Regexp::unbox_from_value(&mut substring, interp) } {
+        let offset_number = match implicitly_convert_to_int(interp, offset.into()) {
+            Err(_) => None,
+            Ok(value) => Some(value),
+        };
```

Comment:
calling `into()` on an `Option<Value>` flattens the option by turning the `None` case into a `nil` Ruby value. Further converting the `Err` return from `implicitly_convert_to_int` squashes this error handling and other errors that may be encountered in the implicit conversion (which can route back through Ruby code to call the `#to_int` method).

This statement doesn't properly handle these error conditions:

```console
$ irb
[3.2.2] > "foo".byteindex(/o/, nil)
(irb):1:in `byteindex': no implicit conversion from nil to integer (TypeError)
        from (irb):1:in `<main>'
        from /usr/local/var/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `<main>'
[3.2.2] > class A; def to_int; 'abc' end; end
=> :to_int
[3.2.2] > "foo".byteindex(/o/, A.new)
(irb):3:in `byteindex': can't convert A to Integer (A#to_int gives String) (TypeError)
        from (irb):3:in `<main>'
        from /usr/local/var/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `<main>'
[3.2.2] > class B; def to_int; raise 'oh no' end; end
=> :to_int
[3.2.2] > "foo".byteindex(/o/, B.new)
(irb):4:in `to_int': oh no (RuntimeError)
        from (irb):5:in `byteindex'
        from (irb):5:in `<main>'
        from /usr/local/var/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `<main>'
```

When calling `implicitly_convert_to_int`, we must use the `?` operator to return errors.

What do you think about replacing this statement with:

```suggestion
        let offset_number = if let Some(offset) = offset {
            Some(implicitly_convert_to_int(interp, offset)?)
        } else {
            None
        };
```

Reply:
Thanks! I've added fixes for all points in 24e032e35764a16d8e6d657d786d79b919e4be38, I can squash the commits later if you prefer. Thanks for the feedback, it was very valuable to me!